<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>_systemPipeR_: NGS workflow and report generation environment  | Manuals</title>
<link rel="stylesheet" href="/manuals/css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="/manuals/css/modern-business.css">
<link rel="stylesheet" href="/manuals/css/lavish-bootstrap.css">
<link rel="stylesheet" href="/manuals/css/customstyles.css">
<link rel="stylesheet" href="/manuals/css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="/manuals/js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="/manuals/js/toc.js"></script>
<script src="/manuals/js/customscripts.js"></script>

<link rel="shortcut icon" href="/manuals/images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="http://idratherbewriting.com/manuals/feed.xml">




 
    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="/manuals/">&nbsp;<span class="projectTitle"> Data Analysis Tutorials</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="/manuals/news/">News</a></li>
                
                
                
                <li><a href="/manuals/news/">News</a></li>
                
                
                
                <li><a href="/manuals/news/">News</a></li>
                
                
                
                <li><a href="/manuals/news/">News</a></li>
                
                
                
                <li><a href="/manuals/news/">News</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Jekyll Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://github.com/tomjohnson1492/documentation-theme-jekyll" target="_blank">Github Repo</a></li>
                        
                        
                        
                        <li><a href="https://talk.jekyllrb.com" target="_blank">Jekyll Talk</a></li>
                        
                        
                        
                        <li><a href="http://jekyllrb.com/docs/home/" target="_blank">Jekyll documentation</a></li>
                        
                        
                        
                        <li><a href="http://stackoverflow.com/questions/tagged/jekyll" target="_blank">Jekyll on Stack Overflow</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/category-jekyll/" target="_blank">Jekyll on my blog</a></li>
                        
                        
                    </ul>
                </li>
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Products<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="/manuals/mydoc_introduction/">Jekyll Documentation Theme</a></li>
                        
                        
                        
                        <li><a href="/manuals/p1_landing_page/">Product 1</a></li>
                        
                        
                        
                        <li><a href="/manuals/p2_landing_page/">Product 2</a></li>
                        
                        
                        
                        <li><a href="/manuals/home/">Manuals</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "Manuals feedback";
var body = "I have some feedback about the _systemPipeR_: NGS workflow and report generation environment page: ";
body += window.location.href;
body += "";
var uri = "mailto:tomjohnson1492@gmail.com?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="/manuals/js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: '/manuals/search.json',
                                searchResultTemplate: '<li><a href="{url}" title="_systemPipeR_: NGS workflow and report generation environment">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>
<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3">

          














<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"> </li>
    
    
    
        
    
    <li>
        <a href="#">Overview</a>
        <ul>
            
            
            
            <li><a href="/manuals/home/">Introduction</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_classes/">Class teaching</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Introduction to R</a>
        <ul>
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_1/">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_2/">2. R Package Repositories</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_3/">3. Installation of R and Add-on Packages</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_4/">4. Getting Around</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_5/">5. Basic Syntax</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_6/">6. Data Types</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_7/">7. Data objects</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_8/">8. Graphics example</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_Rbasics_9/">9. Session Info</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChemmineR</a>
        <ul>
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_01/">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_02/">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_03/">3. OpenBabel Functions</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_04/">4. Overview of Classes and Functions</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_05/">5. Import of Compounds</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_06/">6. Export of Compounds</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_07/">7. Format Interconversions</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_08/">8. Splitting SD Files</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_09/">9. Streaming Through Large SD Files</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_10/">10. Storing Compounds in an SQL Database</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_11/">11. Working with SDF/SDFset Classes</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_12/">12. Molecular Property Functions (Physicochemical Descriptors)</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_13/">13. Bond Matrices</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_14/">14. Charges and Missing Hydrogens</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_15/">15. Ring Perception and Aromaticity Assignment</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_16/">16. Rendering Chemical Structure Images</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_17/">17. Similarity Comparisons and Searching</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_18/">18. Clustering</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_19/">19. Searching PubChem</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_20/">20. ChemMine Tools R Interface</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_21/">21. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_22/">22. Funding</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_ChemmineR_23/">23. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">bioassayR</a>
        <ul>
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_1/">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_2/">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_3/">3. Examples</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_4/">4. Building PubChem BioAssay Database</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_5/">5. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_6/">6. Funding</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_bioassayR_7/">7. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">systemPipeR</a>
        <ul>
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_1/">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_2/">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_3/">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_4/">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_5/">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="/manuals/manuals/mydoc_systemPipeR_6/">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Internal notes</a>
        <ul>
            
            
            
            <li><a href="/manuals/manuals/mydoc_internal_notes/">Administration of site</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>
    <!-- Content Column -->
    <div class="col-md-9">
        <div class="post-header">
   <h1 class="post-title-main">_systemPipeR_: NGS workflow and report generation environment</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    

    

    

    <a target="_blank" href="https://github.com/tomjohnson1492/documentation-theme-jekyll/blob/gh-pages//vignettes07_systemPipeRsystemPipeR.utf8.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    

    

  <!--
%% \VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Overview Vignette}
%% \VignetteDepends{rjson, ggplot2, limma, edgeR, DESeq2, GOstats, GO.db, annotate, pheatmap}
%% \VignetteKeywords{compute cluster, pipeline, reports}
%% \VignettePackage{systemPipeR}
-->

<!---
- Compile from command-line
echo "rmarkdown::render('systemPipeR.Rmd')" | R -slave; R CMD Stangle systemPipeR.Rmd

- Commit to github
git commit -am "some edits"; git push -u origin master

- To customize font size and other style features, add this line to output section in preamble:  
    css: style.css
-->

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'ggplot2::unit' when loading 'systemPipeR'
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'ggplot2::arrow' when loading 'systemPipeR'
</code></pre>
</div>
<p>Note: the most recent version of this tutorial can be found <a href="https://htmlpreview.github.io/?https://github.com/tgirke/systemPipeR/blob/master/vignettes/systemPipeR.html">here</a> and a short overview slide show <a href="https://htmlpreview.github.io/?https://github.com/tgirke/systemPipeR/master/inst/extdata/slides/systemPipeRslides.html">here</a>.</p>

<h1 id="introduction">Introduction</h1>

<p><a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code class="highlighter-rouge">systemPipeR</code></em></a> provides utilities for building and running automated end-to-end analysis workflows for a wide range of next generation sequence (NGS) applications such as RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq [@Girke2014-oy]. Important features include a uniform workflow interface across different NGS applications, automated report generation, and support for running both R and command-line software, such as NGS aligners or peak/variant callers, on local computers or compute clusters. The latter supports interactive job submissions and batch submissions to queuing systems of clusters. For instance, <em><code class="highlighter-rouge">systemPipeR</code></em> can be used with most command-line aligners such as <code class="highlighter-rouge">BWA</code> [@Li2013-oy; @Li2009-oc], <code class="highlighter-rouge">TopHat2</code> [@Kim2013-vg] and <code class="highlighter-rouge">Bowtie2</code> [@Langmead2012-bs], as well as the R-based NGS aligners <a href="http://www.bioconductor.org/packages/devel/bioc/html/Rsubread.html"><em><code class="highlighter-rouge">Rsubread</code></em></a> [@Liao2013-bn] and <a href="http://www.bioconductor.org/packages/devel/bioc/html/gmapR.html"><em><code class="highlighter-rouge">gsnap (gmapR)</code></em></a> [@Wu2010-iq]. Efficient handling of complex sample sets (<em>e.g.</em> FASTQ/BAM files) and experimental designs is facilitated by a well-defined sample annotation infrastructure which improves reproducibility and user-friendliness of many typical analysis workflows in the NGS area [@Lawrence2013-kt].</p>

<p>Motivation and advantages of <em><code class="highlighter-rouge">sytemPipeR</code></em> environment:</p>

<ol>
  <li>Facilitates design of complex NGS workflows involving multiple R/Bioconductor packages</li>
  <li>Common workflow interface for different NGS applications</li>
  <li>Makes NGS analysis with Bioconductor utilities more accessible to new users</li>
  <li>Simplifies usage of command-line software from within R</li>
  <li>Reduces complexity of using compute clusters for R and command-line software</li>
  <li>Accelerates runtime of workflows via parallelzation on computer systems with mutiple CPU cores and/or multiple compute nodes</li>
  <li>Automates generation of analysis reports to improve reproducibility</li>
</ol>

<p>A central concept for designing workflows within the <em><code class="highlighter-rouge">sytemPipeR</code></em> environment is the use of workflow management containers called <em><code class="highlighter-rouge">SYSargs</code></em> (see Figure 1). Instances of this S4 object class are constructed by the <em><code class="highlighter-rouge">systemArgs</code></em> function from two simple tabular files: a <em><code class="highlighter-rouge">targets</code></em> file and a <em><code class="highlighter-rouge">param</code></em> file. The latter is optional for workflow steps lacking command-line software. Typically, a <em><code class="highlighter-rouge">SYSargs</code></em> instance stores all sample-level inputs as well as the paths to the corresponding outputs generated by command-line- or R-based software generating sample-level output files, such as read preprocessors (trimmed/filtered FASTQ files), aligners (SAM/BAM files), variant callers (VCF/BCF files) or peak callers (BED/WIG files). Each sample level input/outfile operation uses its own <em><code class="highlighter-rouge">SYSargs</code></em> instance. The outpaths of <em><code class="highlighter-rouge">SYSargs</code></em> usually define the sample inputs for the next <em><code class="highlighter-rouge">SYSargs</code></em> instance. This connectivity is established by writing the outpaths with the <em><code class="highlighter-rouge">writeTargetsout</code></em> function to a new <em><code class="highlighter-rouge">targets</code></em> file that serves as input to the next <em><code class="highlighter-rouge">systemArgs</code></em> call. Typically, the user has to provide only the initial <em><code class="highlighter-rouge">targets</code></em> file. All downstream <em><code class="highlighter-rouge">targets</code></em> files are generated automatically. By chaining several <em><code class="highlighter-rouge">SYSargs</code></em> steps together one can construct complex workflows involving many sample-level input/output file operations with any combinaton of command-line or R-based software.</p>

<p><img src="SystemPipeR_Workflow.png" alt="" /></p>
<div align="center">**Figure 1:** Workflow design structure of *systemPipeR* </div>
<p>&lt;/br&gt;</p>

<p>The intended way of running <em><code class="highlighter-rouge">sytemPipeR</code></em> workflows is via <em><code class="highlighter-rouge">*.Rnw</code></em> or <em><code class="highlighter-rouge">*.Rmd</code></em> files, which can be executed either line-wise in interactive mode or with a single command from R or the command-line using a <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/Makefile"><em><code class="highlighter-rouge">Makefile</code></em></a>. This way comprehensive and reproducible analysis reports in PDF or HTML format can be generated in a fully automated manner by making use of the highly functional reporting utilities available for R. Templates for setting up custom project reports are provided as <em><code class="highlighter-rouge">*.Rnw</code></em> files by the helper package <em><code class="highlighter-rouge">systemPipeRdata</code></em> and in the vignettes subdirectory of <em><code class="highlighter-rouge">systemPipeR</code></em>. The corresponding PDFs of these report templates are available here: <a href="https://github.com/tgirke/systemPipeR/blob/master/vignettes/systemPipeRNAseq.pdf?raw=true"><em><code class="highlighter-rouge">systemPipeRNAseq</code></em></a>, <a href="https://github.com/tgirke/systemPipeR/blob/master/vignettes/systemPipeRIBOseq.pdf?raw=true"><em><code class="highlighter-rouge">systemPipeRIBOseq</code></em></a>, <a href="https://github.com/tgirke/systemPipeR/blob/master/vignettes/systemPipeChIPseq.pdf?raw=true"><em><code class="highlighter-rouge">systemPipeChIPseq</code></em></a> and <a href="https://github.com/tgirke/systemPipeR/blob/master/vignettes/systemPipeVARseq.pdf?raw=true"><em><code class="highlighter-rouge">systemPipeVARseq</code></em></a>. To work with <em><code class="highlighter-rouge">*.Rnw</code></em> or <em><code class="highlighter-rouge">*.Rmd</code></em> files efficiently, basic knowledge of <a href="https://www.stat.uni-muenchen.de/~leisch/Sweave/"><em><code class="highlighter-rouge">Sweave</code></em></a> or <a href="http://yihui.name/knitr/"><em><code class="highlighter-rouge">knitr</code></em></a> and <a href="http://www.latex-project.org/"><em><code class="highlighter-rouge">Latex</code></em></a> or <a href="http://rmarkdown.rstudio.com/"><em><code class="highlighter-rouge">R Markdown v2</code></em></a> is required.</p>

<div align="right">[Back to Table of Contents]()</div>

<h1 id="getting-started">Getting Started</h1>

<h2 id="installation">Installation</h2>

<p>The R software for running <a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html"><em><code class="highlighter-rouge">systemPipeR</code></em></a> can be downloaded from <a href="http://cran.at.r-project.org/"><em>CRAN</em></a>. The <em><code class="highlighter-rouge">systemPipeR</code></em> environment can be installed from R using the <em><code class="highlighter-rouge">biocLite</code></em> install command. The associated data package <a href="https://github.com/tgirke/systemPipeRdata"><em><code class="highlighter-rouge">systemPipeRdata</code></em></a> can be used to generate <em><code class="highlighter-rouge">systemPipeR</code></em> workflow environments with a single command (<a href="#load-sample-data-and-workflow-templates">see below</a>) containing all parameter files and sample data required to quickly test and run workflows.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">source</span><span class="p">(</span><span class="s2">"http://bioconductor.org/biocLite.R"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Sources the biocLite.R installation script 
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Installs systemPipeR 
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"systemPipeRdata"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Installs systemPipeRdata
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="loading-package-and-documentation">Loading package and documentation</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Loads the package
</span><span class="n">library</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Lists package info
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Opens vignette
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="load-sample-data-and-workflow-templates">Load sample data and workflow templates</h2>
<p>The mini sample FASTQ files used by this overview vignette as well as the associated workflow reporting vignettes can be loaded via the <em><code class="highlighter-rouge">systemPipeRdata</code></em> package as shown below. The chosen data set <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938"><code class="highlighter-rouge">SRP010938</code></a> contains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em> [@Howard2013-fq]. To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the <em>A. thalina</em> genome. The corresponding reference genome sequence (FASTA) and its GFF annotion files (provided in the same download) have been truncated accordingly. This way the entire test sample data set requires less than 200MB disk storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single end) reads or PE reads.</p>

<p>The following loads one of the available NGS workflow templates (here RNA-Seq) into the user’s current working directory. At the moment, the package includes workflow templates for RNA-Seq, ChIP-Seq, VAR-Seq and Ribo-Seq. Templates for additional NGS applications will be provided in the future.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The working environment of the sample data loaded in previous step contains the following preconfigured directory structure:</p>

<ul>
  <li><span style="color:grey"><em><strong>workflow/</strong></em></span>
    <ul>
      <li>This is the directory of the R session running the workflow.</li>
      <li>Run script ( <em>*.Rnw</em> or <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
      <li>Note, this directory can have any name (<em>e.g.</em> <span style="color:grey"><em><strong>rnaseq</strong></em></span>, <span style="color:grey"><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
      <li>Important subdirectories:
        <ul>
          <li><span style="color:grey"><em><strong>param/</strong></em></span>
            <ul>
              <li>Stores parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*_run.sh</em>.</li>
            </ul>
          </li>
          <li><span style="color:grey"><em><strong>data/</strong></em> </span>
            <ul>
              <li>FASTQ samples</li>
              <li>Reference FASTA file</li>
              <li>Annotations</li>
              <li>etc.</li>
            </ul>
          </li>
          <li><span style="color:grey"><em><strong>results/</strong></em></span>
            <ul>
              <li>Alignment, variant and peak files (BAM, VCF, BED)</li>
              <li>Tabular result files</li>
              <li>Images and plots</li>
              <li>etc.</li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>

<p>The sample workflows provided by the package are based on the above directory structure, where directory names are indicated in  <span style="color:grey"><em><strong>grey</strong></em></span>. 
Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>

<p>The following parameter files are included in each workflow template:</p>

<ol>
  <li><em><code class="highlighter-rouge">targets.txt</code></em>: initial one provided by user; downstream <em><code class="highlighter-rouge">targets_*.txt</code></em> files are generated automatically</li>
  <li><em><code class="highlighter-rouge">*.param</code></em>: defines parameter for input/output file operations, <em>e.g.</em> <em><code class="highlighter-rouge">trim.param</code></em>, <em><code class="highlighter-rouge">bwa.param</code></em>, <em><code class="highlighter-rouge">vartools.parm</code></em>, …</li>
  <li><em><code class="highlighter-rouge">*_run.sh</code></em>: optional bash script, <em>e.g.</em>: <em><code class="highlighter-rouge">gatk_run.sh</code></em></li>
  <li>Compute cluster environment (skip on single machine):
    <ul>
      <li><em><code class="highlighter-rouge">.BatchJobs</code></em>: defines type of scheduler for <em><code class="highlighter-rouge">BatchJobs</code></em></li>
      <li><em><code class="highlighter-rouge">*.tmpl</code></em>: specifies parameters of scheduler used by a system, <em>e.g.</em> Torque, SGE, StarCluster, Slurm, etc.</li>
    </ul>
  </li>
</ol>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="structure-of-targets-file">Structure of <em><code class="highlighter-rouge">targets</code></em> file</h2>

<p>The <em><code class="highlighter-rouge">targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code class="highlighter-rouge">targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code class="highlighter-rouge">systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single end (SE) reads, the first three columns are mandatory including their column names, while it is four mandatory columns for FASTQ files of PE reads. All subsequent columns are optional and any number of additional columns can be added as needed.</p>

<h3 id="structure-of-targets-file-for-single-end-se-samples">Structure of <em><code class="highlighter-rouge">targets</code></em> file for single end (SE) samples</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
</span><span class="n">targetspath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"targets.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w"> 
</span><span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                    FileName SampleName Factor SampleLong Experiment        Date
## 1  ./data/SRR446027_1.fastq        M1A     M1  Mock.1h.A          1 23-Mar-2012
## 2  ./data/SRR446028_1.fastq        M1B     M1  Mock.1h.B          1 23-Mar-2012
## 3  ./data/SRR446029_1.fastq        A1A     A1   Avr.1h.A          1 23-Mar-2012
## 4  ./data/SRR446030_1.fastq        A1B     A1   Avr.1h.B          1 23-Mar-2012
## 5  ./data/SRR446031_1.fastq        V1A     V1   Vir.1h.A          1 23-Mar-2012
## 6  ./data/SRR446032_1.fastq        V1B     V1   Vir.1h.B          1 23-Mar-2012
## 7  ./data/SRR446033_1.fastq        M6A     M6  Mock.6h.A          1 23-Mar-2012
## 8  ./data/SRR446034_1.fastq        M6B     M6  Mock.6h.B          1 23-Mar-2012
## 9  ./data/SRR446035_1.fastq        A6A     A6   Avr.6h.A          1 23-Mar-2012
## 10 ./data/SRR446036_1.fastq        A6B     A6   Avr.6h.B          1 23-Mar-2012
## 11 ./data/SRR446037_1.fastq        V6A     V6   Vir.6h.A          1 23-Mar-2012
## 12 ./data/SRR446038_1.fastq        V6B     V6   Vir.6h.B          1 23-Mar-2012
## 13 ./data/SRR446039_1.fastq       M12A    M12 Mock.12h.A          1 23-Mar-2012
## 14 ./data/SRR446040_1.fastq       M12B    M12 Mock.12h.B          1 23-Mar-2012
## 15 ./data/SRR446041_1.fastq       A12A    A12  Avr.12h.A          1 23-Mar-2012
## 16 ./data/SRR446042_1.fastq       A12B    A12  Avr.12h.B          1 23-Mar-2012
## 17 ./data/SRR446043_1.fastq       V12A    V12  Vir.12h.A          1 23-Mar-2012
## 18 ./data/SRR446044_1.fastq       V12B    V12  Vir.12h.B          1 23-Mar-2012
</code></pre>
</div>
<p>To work with custom data, users need to generate a <em><code class="highlighter-rouge">targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code class="highlighter-rouge">targetspath</code></em> the path to the corresponding <em><code class="highlighter-rouge">targets</code></em> file.</p>

<div align="right">[Back to Table of Contents]()</div>

<h3 id="structure-of-targets-file-for-paired-end-pe-samples">Structure of <em><code class="highlighter-rouge">targets</code></em> file for paired end (PE) samples</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">targetspath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"targetsPE.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#"</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                  FileName1                FileName2 SampleName Factor SampleLong Experiment
## 1 ./data/SRR446027_1.fastq ./data/SRR446027_2.fastq        M1A     M1  Mock.1h.A          1
## 2 ./data/SRR446028_1.fastq ./data/SRR446028_2.fastq        M1B     M1  Mock.1h.B          1
</code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="sample-comparisons">Sample comparisons</h3>
<p>Sample comparisons are defined in the header lines of the <em><code class="highlighter-rouge">targets</code></em> file starting with ‘<code class="highlighter-rouge"># &lt;CMP&gt;</code>’.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">readLines</span><span class="p">(</span><span class="n">targetspath</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "# Project ID: Arabidopsis - Pseudomonas alternative splicing study (SRA: SRP010938; PMID: 24098335)"                                                                              
## [2] "# The following line(s) allow to specify the contrasts needed for comparative analyses, such as DEG identification. All possible comparisons can be specified with 'CMPset: ALL'."
## [3] "# &lt;CMP&gt; CMPset1: M1-A1, M1-V1, A1-V1, M6-A6, M6-V6, A6-V6, M12-A12, M12-V12, A12-V12"                                                                                             
## [4] "# &lt;CMP&gt; CMPset2: ALL"
</code></pre>
</div>

<p>The function <em><code class="highlighter-rouge">readComp</code></em> imports the comparison information and stores it in a <em><code class="highlighter-rouge">list</code></em>. Alternatively, <em><code class="highlighter-rouge">readComp</code></em> can obtain the comparison information from the corresponding <em><code class="highlighter-rouge">SYSargs</code></em> object (see below). 
Note, these header lines are optional. They are mainly useful for controlling comparative analyses according to certain biological expectations, such as identifying differentially expressed genes in RNA-Seq experiments based on simple pair-wise comparisons.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">readComp</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="o">=</span><span class="s2">"vector"</span><span class="p">,</span><span class="w"> </span><span class="n">delim</span><span class="o">=</span><span class="s2">"-"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## $CMPset1
## [1] "M1-A1"   "M1-V1"   "A1-V1"   "M6-A6"   "M6-V6"   "A6-V6"   "M12-A12" "M12-V12" "A12-V12"
## 
## $CMPset2
##  [1] "M1-A1"   "M1-V1"   "M1-M6"   "M1-A6"   "M1-V6"   "M1-M12"  "M1-A12"  "M1-V12"  "A1-V1"  
## [10] "A1-M6"   "A1-A6"   "A1-V6"   "A1-M12"  "A1-A12"  "A1-V12"  "V1-M6"   "V1-A6"   "V1-V6"  
## [19] "V1-M12"  "V1-A12"  "V1-V12"  "M6-A6"   "M6-V6"   "M6-M12"  "M6-A12"  "M6-V12"  "A6-V6"  
## [28] "A6-M12"  "A6-A12"  "A6-V12"  "V6-M12"  "V6-A12"  "V6-V12"  "M12-A12" "M12-V12" "A12-V12"
</code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="structure-of-param-file-and-sysargs-container">Structure of <em><code class="highlighter-rouge">param</code></em> file and <em><code class="highlighter-rouge">SYSargs</code></em> container</h2>
<p>The <em><code class="highlighter-rouge">param</code></em> file defines the parameters of a chosen command-line software. The following shows the format of a sample <em><code class="highlighter-rouge">param</code></em> file provided by this package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">parampath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">read.delim</span><span class="p">(</span><span class="n">parampath</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"#"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      PairSet         Name                                  Value
## 1    modules         &lt;NA&gt;                          bowtie2/2.2.5
## 2    modules         &lt;NA&gt;                          tophat/2.0.14
## 3   software         &lt;NA&gt;                                 tophat
## 4      cores           -p                                      4
## 5      other         &lt;NA&gt; -g 1 --segment-length 25 -i 30 -I 3000
## 6   outfile1           -o                            &lt;FileName1&gt;
## 7   outfile1         path                             ./results/
## 8   outfile1       remove                                   &lt;NA&gt;
## 9   outfile1       append                                .tophat
## 10  outfile1 outextension              .tophat/accepted_hits.bam
## 11 reference         &lt;NA&gt;                    ./data/tair10.fasta
## 12   infile1         &lt;NA&gt;                            &lt;FileName1&gt;
## 13   infile1         path                                   &lt;NA&gt;
## 14   infile2         &lt;NA&gt;                            &lt;FileName2&gt;
## 15   infile2         path                                   &lt;NA&gt;
</code></pre>
</div>

<p>The <em><code class="highlighter-rouge">systemArgs</code></em> function imports the definitions of both the <em><code class="highlighter-rouge">param</code></em> file and the <em><code class="highlighter-rouge">targets</code></em> file, and stores all relevant information in a <em><code class="highlighter-rouge">SYSargs</code></em> S4 class object. To run the pipeline without command-line software, one can assign <em><code class="highlighter-rouge">NULL</code></em> to <em><code class="highlighter-rouge">sysma</code></em> instead of a <em><code class="highlighter-rouge">param</code></em> file. In addition, one can start the <em><code class="highlighter-rouge">systemPipeR</code></em> workflow with pre-generated BAM files by providing a targets file where the <em><code class="highlighter-rouge">FileName</code></em> column gives the paths to the BAM files and <em><code class="highlighter-rouge">sysma</code></em> is assigned <em><code class="highlighter-rouge">NULL</code></em>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">suppressWarnings</span><span class="p">(</span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="n">parampath</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="n">targetspath</span><span class="p">))</span><span class="w">
</span><span class="n">args</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## An instance of 'SYSargs' for running 'tophat' on 18 samples
</code></pre>
</div>

<p>Several accessor functions are available that are named after the slot names of the <em><code class="highlighter-rouge">SYSargs</code></em> object.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">names</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] "targetsin"     "targetsout"    "targetsheader" "modules"       "software"      "cores"        
##  [7] "other"         "reference"     "results"       "infile1"       "infile2"       "outfile1"     
## [13] "sysargs"       "outpaths"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "bowtie2/2.2.5" "tophat/2.0.14"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 4
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                                                                                                              M1A 
## "/home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/results/SRR446027_1.fastq.tophat/accepted_hits.bam"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sysargs</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                                                                                                                                                                                                                                                                                  M1A 
## "tophat -p 4 -g 1 --segment-length 25 -i 30 -I 3000 -o /home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/results/SRR446027_1.fastq.tophat /home/tgirke/Dropbox/Websites/manuals/vignettes/systemPipeR/data/tair10.fasta ./data/SRR446027_1.fastq ./data/SRR446027_2.fastq"
</code></pre>
</div>

<p>The content of the <em><code class="highlighter-rouge">param</code></em> file can also be returned as JSON object as follows (requires <em><code class="highlighter-rouge">rjson</code></em> package).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="n">parampath</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"json"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "{\"modules\":{\"n1\":\"\",\"v2\":\"bowtie2/2.2.5\",\"n1\":\"\",\"v2\":\"tophat/2.0.14\"},\"software\":{\"n1\":\"\",\"v1\":\"tophat\"},\"cores\":{\"n1\":\"-p\",\"v1\":\"4\"},\"other\":{\"n1\":\"\",\"v1\":\"-g 1 --segment-length 25 -i 30 -I 3000\"},\"outfile1\":{\"n1\":\"-o\",\"v2\":\"&lt;FileName1&gt;\",\"n3\":\"path\",\"v4\":\"./results/\",\"n5\":\"remove\",\"v1\":\"\",\"n2\":\"append\",\"v3\":\".tophat\",\"n4\":\"outextension\",\"v5\":\".tophat/accepted_hits.bam\"},\"reference\":{\"n1\":\"\",\"v1\":\"./data/tair10.fasta\"},\"infile1\":{\"n1\":\"\",\"v2\":\"&lt;FileName1&gt;\",\"n1\":\"path\",\"v2\":\"\"},\"infile2\":{\"n1\":\"\",\"v2\":\"&lt;FileName2&gt;\",\"n1\":\"path\",\"v2\":\"\"}}"
</code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h1 id="workflow-overview">Workflow overview</h1>

<h2 id="define-environment-settings-and-samples">Define environment settings and samples</h2>

<p>Load packages and generate workflow environment (here for RNA-Seq)</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Construct <em><code class="highlighter-rouge">SYSargs</code></em> object from <em><code class="highlighter-rouge">param</code></em> and <em><code class="highlighter-rouge">targets</code></em> files.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/trim.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="read-preprocessing">Read Preprocessing</h2>
<p>The function <em><code class="highlighter-rouge">preprocessReads</code></em> allows to apply predefined or custom
read preprocessing functions to all FASTQ files referenced in a
<em><code class="highlighter-rouge">SYSargs</code></em> container, such as quality filtering or adaptor trimming
routines. The paths to the resulting output FASTQ files are stored in the
<em><code class="highlighter-rouge">outpaths</code></em> slot of the <em><code class="highlighter-rouge">SYSargs</code></em> object. Internally,
<em><code class="highlighter-rouge">preprocessReads</code></em> uses the <em><code class="highlighter-rouge">FastqStreamer</code></em> function from
the <em><code class="highlighter-rouge">ShortRead</code></em> package to stream through large FASTQ files in a
memory-efficient manner. The following example performs adaptor trimming with
the <em><code class="highlighter-rouge">trimLRPatterns</code></em> function from the <em><code class="highlighter-rouge">Biostrings</code></em> package.
After the trimming step a new targets file is generated (here
<em><code class="highlighter-rouge">targets_trim.txt</code></em>) containing the paths to the trimmed FASTQ files.
The new targets file can be used for the next workflow step with an updated
<em><code class="highlighter-rouge">SYSargs</code></em> instance, \textit{e.g.} running the NGS alignments using the
trimmed FASTQ files.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">preprocessReads</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">Fct</span><span class="o">=</span><span class="s2">"trimLRPatterns(Rpattern='GCCCGGGTAA', subject=fq)"</span><span class="p">,</span><span class="w"> 
                </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">compress</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">writeTargetsout</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"targets_trim.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The following example shows how one can design a custom read preprocessing function 
using utilities provided by the <em><code class="highlighter-rouge">ShortRead</code></em> package, and then run it
in batch mode with the <em>‘preprocessReads’</em> function (here on paired-end reads).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/trimPE.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">filterFct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">fq</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">Nexceptions</span><span class="o">=</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">qcount</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">as</span><span class="p">(</span><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span><span class="w"> </span><span class="s2">"matrix"</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">cutoff</span><span class="p">)</span><span class="w">
    </span><span class="n">fq</span><span class="p">[</span><span class="n">qcount</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">Nexceptions</span><span class="p">]</span><span class="w"> </span><span class="c1"># Retains reads where Phred scores are &gt;= cutoff with N exceptions
</span><span class="p">}</span><span class="w">
</span><span class="n">preprocessReads</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">Fct</span><span class="o">=</span><span class="s2">"filterFct(fq, cutoff=20, Nexceptions=0)"</span><span class="p">,</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">)</span><span class="w">
</span><span class="n">writeTargetsout</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"targets_PEtrim.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="fastq-quality-report">FASTQ quality report</h2>
<p>The following <em><code class="highlighter-rouge">seeFastq</code></em> and <em><code class="highlighter-rouge">seeFastqPlot</code></em> functions generate and plot a series of
useful quality statistics for a set of FASTQ files including per cycle quality
box plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads above
quality cutoffs and mean quality distribution.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">10000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"./results/fastqReport.pdf"</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">4</span><span class="o">*</span><span class="nf">length</span><span class="p">(</span><span class="n">fqlist</span><span class="p">))</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">fqlist</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>
<p><img src="fastqReport.png" alt="" /></p>
<div align="center">**Figure 2:** FASTQ quality report </div>
<p>&lt;/br&gt;</p>

<div align="right">[Back to Table of Contents]()</div>

<p>Parallelization of QC report on single machine with multiple cores</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">8</span><span class="p">))</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">fqlist</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<p>Parallelization of QC report via scheduler (<em>e.g.</em> Torque) across several compute nodes</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsTorque</span><span class="p">(</span><span class="s2">"torque.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="s2">"1:ppn=1"</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">fqlist</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="alignment-with-tophat2">Alignment with <em><code class="highlighter-rouge">Tophat2</code></em></h2>
<p>Build <em><code class="highlighter-rouge">Bowtie2</code></em> index.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available
</span><span class="n">system</span><span class="p">(</span><span class="s2">"bowtie2-build ./data/tair10.fasta ./data/tair10.fasta"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Execute <em><code class="highlighter-rouge">SYSargs</code></em> on a single machine without submitting to a queuing system of a compute cluster. This way the input FASTQ files will be processed sequentially. If available, multiple CPU cores can be used for processing each file. The number of CPU cores (here 4) to use for each process is defined in the <em><code class="highlighter-rouge">*.param</code></em> file. With <em><code class="highlighter-rouge">cores(args)</code></em> one can return this value from the <em><code class="highlighter-rouge">SYSargs</code></em> object. Note, if a module system is not installed or used, then the corresponding <em><code class="highlighter-rouge">*.param</code></em> file needs to be edited accordingly by either providing an empty field in the line(s) starting with <em><code class="highlighter-rouge">module</code></em> or by deleting these lines.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. To avoid over-subscription of CPU cores on the compute nodes, the value from <em><code class="highlighter-rouge">cores(args)</code></em> is passed on to the submission command, here <em><code class="highlighter-rouge">nodes</code></em> in the <em><code class="highlighter-rouge">resources</code></em> list object. The number of independent parallel cluster processes is defined under the <em><code class="highlighter-rouge">Njobs</code></em> argument. The following example will run 18 processes in parallel using for each 4 CPU cores. If the resources available on a cluster allow to run all 18 processes at the same time then the shown sample submission will utilize in total 72 CPU cores. Note, <em><code class="highlighter-rouge">clusterRun</code></em> can be used with most queueing systems as it is based on utilities from the <em><code class="highlighter-rouge">BatchJobs</code></em> package which supports the use of template files (<em><code class="highlighter-rouge">*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a conf file (see <em><code class="highlighter-rouge">.BatchJob</code></em> samples <a href="https://code.google.com/p/batchjobs/wiki/DortmundUsage">here</a>) and a template file (see <em><code class="highlighter-rouge">*.tmpl</code></em> samples <a href="https://github.com/tudo-r/BatchJobs/tree/master/examples">here</a>) for the queueing available on a system. The following example uses the sample conf and template files for the Torque scheduler provided by this package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">resources</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"1:ppn="</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">)),</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"10gb"</span><span class="p">)</span><span class="w">
</span><span class="n">reg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterRun</span><span class="p">(</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">conffile</span><span class="o">=</span><span class="s2">".BatchJobs.R"</span><span class="p">,</span><span class="w"> </span><span class="n">template</span><span class="o">=</span><span class="s2">"torque.tmpl"</span><span class="p">,</span><span class="w"> </span><span class="n">Njobs</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">runid</span><span class="o">=</span><span class="s2">"01"</span><span class="p">,</span><span class="w"> 
                  </span><span class="n">resourceList</span><span class="o">=</span><span class="n">resources</span><span class="p">)</span><span class="w">
</span><span class="n">waitForJobs</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Useful commands for monitoring progress of submitted jobs</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">showStatus</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span><span class="n">file.exists</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w">
</span><span class="n">sapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">loadResult</span><span class="p">(</span><span class="n">reg</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">))</span><span class="w"> </span><span class="c1"># Works after job completion
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="read-and-alignment-count-stats">Read and alignment count stats</h2>
<p>Generate table of read and alignment counts for all samples.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="w"> 
</span><span class="n">write.table</span><span class="p">(</span><span class="n">read_statsDF</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/alignStats.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The following shows the first four lines of the sample alignment stats file provided by the <em><code class="highlighter-rouge">systemPipeR</code></em> package. For simplicity the number of PE reads is multiplied here by 2 to approximate proper alignment frequencies where each read in a pair is counted.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">read.table</span><span class="p">(</span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"alignStats.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   FileName Nreads2x Nalign Perc_Aligned Nalign_Primary Perc_Aligned_Primary
## 1      M1A   192918 177961     92.24697         177961             92.24697
## 2      M1B   197484 159378     80.70426         159378             80.70426
## 3      A1A   189870 176055     92.72397         176055             92.72397
## 4      A1B   188854 147768     78.24457         147768             78.24457
</code></pre>
</div>

<p>Parallelization of read/alignment stats on single machine with multiple cores</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="n">read_statsList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">8</span><span class="p">))</span><span class="w">
</span><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">do.call</span><span class="p">(</span><span class="s2">"rbind"</span><span class="p">,</span><span class="w"> </span><span class="n">read_statsList</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Parallelization of read/alignment stats via scheduler (<em>e.g.</em> Torque) across several compute nodes</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsTorque</span><span class="p">(</span><span class="s2">"torque.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="s2">"1:ppn=1"</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">read_statsList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">do.call</span><span class="p">(</span><span class="s2">"rbind"</span><span class="p">,</span><span class="w"> </span><span class="n">read_statsList</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h2>
<p>The genome browser IGV supports reading of indexed/sorted BAM files via web URLs. This way it can be avoided to create unnecessary copies of these large files. To enable this approach, an HTML directory with http access needs to be available in the user account (<em>e.g.</em> <em><code class="highlighter-rouge">home/publichtml</code></em>) of a system. If this is not the case then the BAM files need to be moved or copied to the system where IGV runs. In the following, <em><code class="highlighter-rouge">htmldir</code></em> defines the path to the HTML directory with http access where the symbolic links to the BAM files will be stored. The corresponding URLs will be written to a text file specified under the <code class="highlighter-rouge">_urlfile</code>_ argument.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">symLink2bam</span><span class="p">(</span><span class="n">sysargs</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">htmldir</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"~/.html/"</span><span class="p">,</span><span class="w"> </span><span class="s2">"somedir/"</span><span class="p">),</span><span class="w"> 
            </span><span class="n">urlbase</span><span class="o">=</span><span class="s2">"http://myserver.edu/~username/"</span><span class="p">,</span><span class="w"> 
        </span><span class="n">urlfile</span><span class="o">=</span><span class="s2">"IGVurl.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="alternative-ngs-aligners">Alternative NGS Aligners</h2>

<h3 id="alignment-with-bowtie2-eg-for-mirna-profiling">Alignment with <em><code class="highlighter-rouge">Bowtie2</code></em> (<em>e.g.</em> for miRNA profiling)</h3>
<p>The following example runs <em><code class="highlighter-rouge">Bowtie2</code></em> as a single process without submitting it to a cluster.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"bowtieSE.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available
</span><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Alternatively, submit the job to compute nodes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">resources</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"1:ppn="</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">)),</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"10gb"</span><span class="p">)</span><span class="w">
</span><span class="n">reg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterRun</span><span class="p">(</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">conffile</span><span class="o">=</span><span class="s2">".BatchJobs.R"</span><span class="p">,</span><span class="w"> </span><span class="n">template</span><span class="o">=</span><span class="s2">"torque.tmpl"</span><span class="p">,</span><span class="w"> </span><span class="n">Njobs</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">runid</span><span class="o">=</span><span class="s2">"01"</span><span class="p">,</span><span class="w"> 
                  </span><span class="n">resourceList</span><span class="o">=</span><span class="n">resources</span><span class="p">)</span><span class="w">
</span><span class="n">waitForJobs</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="alignment-with-bwa-mem-eg-for-var-seq">Alignment with <em><code class="highlighter-rouge">BWA-MEM</code></em> (<em>e.g.</em> for VAR-Seq)</h3>
<p>The following example runs BWA-MEM as a single process without submitting it to a cluster.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/bwa.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available
</span><span class="n">system</span><span class="p">(</span><span class="s2">"bwa index -a bwtsw ./data/tair10.fasta"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Indexes reference genome
</span><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="alignment-with-rsubread-eg-for-rna-seq">Alignment with <em><code class="highlighter-rouge">Rsubread</code></em> (<em>e.g.</em> for RNA-Seq)</h3>
<p>The following example shows how one can use within the \Rpackage{systemPipeR} environment the R-based 
aligner \Rpackage{Rsubread} or other R-based functions that read from input files and write to output files.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Rsubread</span><span class="p">)</span><span class="w">
</span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/rsubread.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">buildindex</span><span class="p">(</span><span class="n">basename</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Build indexed reference genome
</span><span class="n">align</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">readfile1</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">input_format</span><span class="o">=</span><span class="s2">"FASTQ"</span><span class="p">,</span><span class="w"> 
      </span><span class="n">output_file</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">output_format</span><span class="o">=</span><span class="s2">"SAM"</span><span class="p">,</span><span class="w"> </span><span class="n">nthreads</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">indels</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">TH1</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)))</span><span class="w"> </span><span class="n">asBam</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">destination</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">".sam"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">i</span><span class="p">]),</span><span class="w"> </span><span class="n">overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">indexDestination</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="alignment-with-gsnap-eg-for-var-seq-and-rna-seq">Alignment with <em><code class="highlighter-rouge">gsnap</code></em> (<em>e.g.</em> for VAR-Seq and RNA-Seq)</h3>
<p>Another R-based short read aligner is <em><code class="highlighter-rouge">gsnap</code></em> from the <em><code class="highlighter-rouge">gmapR</code></em> package [@Wu2010-iq]. 
The code sample below introduces how to run this aligner on multiple nodes of a compute cluster.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gmapR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/gsnap.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">gmapGenome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GmapGenome</span><span class="p">(</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">directory</span><span class="o">=</span><span class="s2">"data"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="o">=</span><span class="s2">"gmap_tair10chr/"</span><span class="p">,</span><span class="w"> </span><span class="n">create</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">gmapR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"gsnap.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">gmapGenome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GmapGenome</span><span class="p">(</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">directory</span><span class="o">=</span><span class="s2">"data"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="o">=</span><span class="s2">"gmap_tair10chr/"</span><span class="p">,</span><span class="w"> </span><span class="n">create</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
    </span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GsnapParam</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="n">gmapGenome</span><span class="p">,</span><span class="w"> </span><span class="n">unique_only</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">molecule</span><span class="o">=</span><span class="s2">"DNA"</span><span class="p">,</span><span class="w"> </span><span class="n">max_mismatches</span><span class="o">=</span><span class="m">3</span><span class="p">)</span><span class="w">
    </span><span class="n">o</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsnap</span><span class="p">(</span><span class="n">input_a</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">input_b</span><span class="o">=</span><span class="n">infile2</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">params</span><span class="o">=</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">output</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsTorque</span><span class="p">(</span><span class="s2">"torque.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="s2">"1:ppn=1"</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="read-counting-for-mrna-profiling-experiments">Read counting for mRNA profiling experiments</h2>
<p>Create <em><code class="highlighter-rouge">txdb</code></em> (needs to be done only once)</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">)</span><span class="w">
</span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeTxDbFromGFF</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"data/tair10.gff"</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="o">=</span><span class="s2">"gff"</span><span class="p">,</span><span class="w"> </span><span class="n">dataSource</span><span class="o">=</span><span class="s2">"TAIR"</span><span class="p">,</span><span class="w"> </span><span class="n">organism</span><span class="o">=</span><span class="s2">"A. thaliana"</span><span class="p">)</span><span class="w">
</span><span class="n">saveDb</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<p>The following performs read counting with <em><code class="highlighter-rouge">summarizeOverlaps</code></em> in parallel mode with multiple cores.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span><span class="w">
</span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">loadDb</span><span class="p">(</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
</span><span class="n">eByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">)</span><span class="w">
</span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
</span><span class="n">multicoreParam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">4</span><span class="p">);</span><span class="w"> </span><span class="n">register</span><span class="p">(</span><span class="n">multicoreParam</span><span class="p">);</span><span class="w"> </span><span class="n">registered</span><span class="p">()</span><span class="w">
</span><span class="n">counteByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">bfl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">eByg</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">singleEnd</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="c1"># Note: for strand-specific RNA-Seq set 'ignore.strand=FALSE' and for PE data set 'singleEnd=FALSE'
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">counteByg</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">assays</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span><span class="o">$</span><span class="n">counts</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">rowRanges</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="m">1</span><span class="p">]]));</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bfl</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">returnRPKM</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">ranges</span><span class="o">=</span><span class="n">eByg</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/rpkmDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<p>Please note, in addition to read counts this step generates RPKM normalized expression values. For most statistical differential expression or abundance analysis methods, such as <em><code class="highlighter-rouge">edgeR</code></em> or <em><code class="highlighter-rouge">DESeq2</code></em>, the raw count values should be used as input. The usage of RPKM values should be restricted to specialty applications required by some users, <em>e.g.</em> manually comparing the expression levels of different genes or features.</p>

<div align="right">[Back to Table of Contents]()</div>

<p>Read counting with <em><code class="highlighter-rouge">summarizeOverlaps</code></em> using multiple nodes of a cluster</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">)</span><span class="w">
    </span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">loadDb</span><span class="p">(</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
    </span><span class="n">eByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
    </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">eByg</span><span class="p">,</span><span class="w"> </span><span class="n">bfl</span><span class="p">[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">singleEnd</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsTorque</span><span class="p">(</span><span class="s2">"torque.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">nodes</span><span class="o">=</span><span class="s2">"1:ppn=1"</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">counteByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w"> 
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">counteByg</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">assays</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span><span class="o">$</span><span class="n">counts</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">rowRanges</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="m">1</span><span class="p">]]));</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="read-counting-for-mirna-profiling-experiments">Read counting for miRNA profiling experiments</h2>
<p>Download miRNA genes from miRBase</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">system</span><span class="p">(</span><span class="s2">"wget ftp://mirbase.org/pub/mirbase/19/genomes/My_species.gff3 -P ./data/"</span><span class="p">)</span><span class="w">
</span><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">import.gff</span><span class="p">(</span><span class="s2">"./data/My_species.gff3"</span><span class="p">)</span><span class="w">
</span><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">elementMetadata</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">$</span><span class="n">ID</span><span class="p">)</span><span class="w">
</span><span class="n">bams</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bampaths</span><span class="p">);</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bams</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">targets</span><span class="o">$</span><span class="n">SampleName</span><span class="w">
</span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">bams</span><span class="p">,</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
</span><span class="n">countDFmiR</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">bfl</span><span class="p">,</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="c1"># Note: inter.feature=FALSE important since pre and mature miRNA ranges overlap
</span><span class="n">rpkmDFmiR</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">countDFmiR</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">returnRPKM</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">gffsub</span><span class="o">=</span><span class="n">gff</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">assays</span><span class="p">(</span><span class="n">countDFmiR</span><span class="p">)</span><span class="o">$</span><span class="n">counts</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/countDFmiR.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">rpkmDFmiR</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/rpkmDFmiR.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="correlation-analysis-of-samples">Correlation analysis of samples</h2>
<p>The following computes the sample-wise Spearman correlation coefficients from the <em><code class="highlighter-rouge">rlog</code></em> (regularized-logarithm) transformed expression values generated with the <em><code class="highlighter-rouge">DESeq2</code></em> package. After transformation to a distance matrix, hierarchical clustering is performed with the <em><code class="highlighter-rouge">hclust</code></em> function and the result is plotted as a dendrogram (<a href="./results/sample_tree.pdf">sample_tree.pdf</a>).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">,</span><span class="w"> </span><span class="n">warn.conflicts</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">quietly</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">,</span><span class="w"> </span><span class="n">warn.conflicts</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'ggplot2::Position' when loading 'DESeq2'
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'ggplot2::unit' when loading 'Hmisc'
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'ggplot2::arrow' when loading 'Hmisc'
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: replacing previous import by 'scales::alpha' when loading 'Hmisc'
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">countDFpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">countDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">read.table</span><span class="p">(</span><span class="n">countDFpath</span><span class="p">))</span><span class="w">
</span><span class="n">colData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="o">=</span><span class="n">targetsin</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">$</span><span class="n">SampleName</span><span class="p">,</span><span class="w"> </span><span class="n">condition</span><span class="o">=</span><span class="n">targetsin</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">$</span><span class="n">Factor</span><span class="p">)</span><span class="w">
</span><span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">countDF</span><span class="p">,</span><span class="w"> </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colData</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">condition</span><span class="p">)</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">assay</span><span class="p">(</span><span class="n">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"spearman"</span><span class="p">)</span><span class="w">
</span><span class="n">hc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">dist</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">d</span><span class="p">))</span><span class="w">
</span><span class="n">plot.phylo</span><span class="p">(</span><span class="n">as.phylo</span><span class="p">(</span><span class="n">hc</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.col</span><span class="o">=</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">edge.width</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">show.node.label</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">no.margin</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="systemPipeR_files/figure-html/sample_tree_rlog-1.png" alt="" />\</p>

<div align="center">**Figure 3:** Correlation dendrogram of samples for _`rlog`_ values. </div>
<p>&lt;/br&gt;</p>

<p>Alternatively, the clustering can be performed with <em><code class="highlighter-rouge">RPKM</code></em> normalized expression values. In combination with Spearman correlation the results of the two clustering methods are often relatively similar.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rpkmDFeBygpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"rpkmDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">rpkmDFeBygpath</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rpkmDFeByg</span><span class="p">[</span><span class="n">rowMeans</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">,]</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"spearman"</span><span class="p">)</span><span class="w">
</span><span class="n">hc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">as.dist</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">d</span><span class="p">))</span><span class="w">
</span><span class="n">plot.phylo</span><span class="p">(</span><span class="n">as.phylo</span><span class="p">(</span><span class="n">hc</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.col</span><span class="o">=</span><span class="s2">"blue"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.width</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">show.node.label</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">no.margin</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="deg-analysis-with-edger">DEG analysis with <em><code class="highlighter-rouge">edgeR</code></em></h2>
<p>The following <em><code class="highlighter-rouge">run_edgeR</code></em> function is a convenience wrapper for
identifying differentially expressed genes (DEGs) in batch mode with
<em><code class="highlighter-rouge">edgeR</code></em>’s GML method [@Robinson2010-uk] for any number of
pairwise sample comparisons specified under the <em><code class="highlighter-rouge">cmp</code></em> argument. Users
are strongly encouraged to consult the 
<a href="\href{http://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf"><em><code class="highlighter-rouge">edgeR</code></em></a> vignette 
for more detailed information on this topic and how to properly run <em><code class="highlighter-rouge">edgeR</code></em> 
on data sets with more complex experimental designs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">targets</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">comment</span><span class="o">=</span><span class="s2">"#"</span><span class="p">)</span><span class="w">
</span><span class="n">cmp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readComp</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="o">=</span><span class="s2">"matrix"</span><span class="p">,</span><span class="w"> </span><span class="n">delim</span><span class="o">=</span><span class="s2">"-"</span><span class="p">)</span><span class="w">
</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##       [,1]  [,2] 
##  [1,] "M1"  "A1" 
##  [2,] "M1"  "V1" 
##  [3,] "A1"  "V1" 
##  [4,] "M6"  "A6" 
##  [5,] "M6"  "V6" 
##  [6,] "A6"  "V6" 
##  [7,] "M12" "A12"
##  [8,] "M12" "V12"
##  [9,] "A12" "V12"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">countDFeBygpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">countDFeBygpath</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="n">edgeDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">run_edgeR</span><span class="p">(</span><span class="n">countDF</span><span class="o">=</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">targets</span><span class="o">=</span><span class="n">targets</span><span class="p">,</span><span class="w"> </span><span class="n">cmp</span><span class="o">=</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span><span class="w"> </span><span class="n">independent</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">mdsplot</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Disp = 0.20653 , BCV = 0.4545
</code></pre>
</div>

<p>Filter and plot DEG results for up and down regulated genes. Because of the small size of the toy data set used by this vignette, the <em>FDR</em> value has been set to a relatively high threshold (here 10%). More commonly used <em>FDR</em> cutoffs are 1% or 5%. The definition of ‘<em>up</em>’ and ‘<em>down</em>’ is given in the corresponding help file. To open it, type <em><code class="highlighter-rouge">?filterDEGs</code></em> in the R console.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEG_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">edgeDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">10</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="systemPipeR_files/figure-html/edger_deg_counts-1.png" alt="" />\</p>
<div align="center">**Figure 4:** Up and down regulated DEGs identified by _`edgeR`_. </div>
<p>&lt;/br&gt;</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">names</span><span class="p">(</span><span class="n">DEG_list</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "UporDown" "Up"       "Down"     "Summary"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEG_list</span><span class="o">$</span><span class="n">Summary</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##       Comparisons Counts_Up_or_Down Counts_Up Counts_Down
## M1-A1       M1-A1                 0         0           0
## M1-V1       M1-V1                 1         1           0
## A1-V1       A1-V1                 1         1           0
## M6-A6       M6-A6                 0         0           0
</code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h2 id="deg-analysis-with-deseq2">DEG analysis with <em><code class="highlighter-rouge">DESeq2</code></em></h2>
<p>The following <em><code class="highlighter-rouge">run_DESeq2</code></em> function is a convenience wrapper for
identifying DEGs in batch mode with <em><code class="highlighter-rouge">DESeq2</code></em> [@Love2014-sh] for any number of
pairwise sample comparisons specified under the <em><code class="highlighter-rouge">cmp</code></em> argument. Users
are strongly encouraged to consult the 
<a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf"><em><code class="highlighter-rouge">DESeq2</code></em></a> vignette
for more detailed information on this topic and how to properly run <em><code class="highlighter-rouge">DESeq2</code></em> 
on data sets with more complex experimental designs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">degseqDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">run_DESeq2</span><span class="p">(</span><span class="n">countDF</span><span class="o">=</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">targets</span><span class="o">=</span><span class="n">targets</span><span class="p">,</span><span class="w"> </span><span class="n">cmp</span><span class="o">=</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span><span class="w"> </span><span class="n">independent</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>Filter and plot DEG results for up and down regulated genes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEG_list2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">degseqDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">10</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="systemPipeR_files/figure-html/deseq2_deg_counts-1.png" alt="" />\</p>
<div align="center">**Figure 5:** Up and down regulated DEGs identified by _`DESeq2`_. </div>
<p>&lt;/br&gt;</p>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="venn-diagrams">Venn Diagrams</h2>
<p>The function <em><code class="highlighter-rouge">overLapper</code></em> can compute Venn intersects for large numbers of sample sets (up to 20 or more) and <em><code class="highlighter-rouge">vennPlot</code></em> can plot 2-5 way Venn diagrams. A useful feature is the possiblity to combine the counts from several Venn comparisons with the same number of sample sets in a single Venn diagram (here for 4 up and down DEG sets).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">vennsetup</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">overLapper</span><span class="p">(</span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Up</span><span class="p">[</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"vennsets"</span><span class="p">)</span><span class="w">
</span><span class="n">vennsetdown</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">overLapper</span><span class="p">(</span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Down</span><span class="p">[</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"vennsets"</span><span class="p">)</span><span class="w">
</span><span class="n">vennPlot</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">vennsetup</span><span class="p">,</span><span class="w"> </span><span class="n">vennsetdown</span><span class="p">),</span><span class="w"> </span><span class="n">mymain</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">mysub</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">colmode</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ccol</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"blue"</span><span class="p">,</span><span class="w"> </span><span class="s2">"red"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="systemPipeR_files/figure-html/vennplot-1.png" alt="" />\</p>
<div align="center">**Figure 6:** Venn Diagram for 4 Up and Down DEG Sets. </div>
<p>&lt;/br&gt;</p>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="go-term-enrichment-analysis-of-degs">GO term enrichment analysis of DEGs</h2>
<p>### Obtain gene-to-GO mappings
The following shows how to obtain gene-to-GO mappings from <em><code class="highlighter-rouge">biomaRt</code></em> (here for <em>A. thaliana</em>) and how to organize them for the downstream GO term enrichment analysis. Alternatively, the gene-to-GO mappings can be obtained for many organisms from Bioconductor’s  <em><code class="highlighter-rouge">*.db</code></em> genome annotation packages or GO annotation files provided by various genome databases. For each annotation this relatively slow preprocessing step needs to be performed only once. Subsequently, the preprocessed data can be loaded with the <em><code class="highlighter-rouge">load</code></em> function as shown in the next subsection.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"biomaRt"</span><span class="p">)</span><span class="w">
</span><span class="n">listMarts</span><span class="p">()</span><span class="w"> </span><span class="c1"># To choose BioMart database
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">);</span><span class="w"> </span><span class="n">listDatasets</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> 
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="o">=</span><span class="s2">"athaliana_eg_gene"</span><span class="p">)</span><span class="w">
</span><span class="n">listAttributes</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="c1"># Choose data types you want to download
</span><span class="n">go</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"go_accession"</span><span class="p">,</span><span class="w"> </span><span class="s2">"tair_locus"</span><span class="p">,</span><span class="w"> </span><span class="s2">"go_namespace_1003"</span><span class="p">),</span><span class="w"> </span><span class="n">mart</span><span class="o">=</span><span class="n">m</span><span class="p">)</span><span class="w">
</span><span class="n">go</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">go</span><span class="p">[</span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="o">!=</span><span class="s2">""</span><span class="p">,];</span><span class="w"> </span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">])</span><span class="w">
</span><span class="n">dir.create</span><span class="p">(</span><span class="s2">"./data/GO"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">go</span><span class="p">,</span><span class="w"> </span><span class="s2">"data/GO/GOannotationsBiomart_mod.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">catdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeCATdb</span><span class="p">(</span><span class="n">myfile</span><span class="o">=</span><span class="s2">"data/GO/GOannotationsBiomart_mod.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">lib</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">org</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">colno</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="n">idconv</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">save</span><span class="p">(</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"data/GO/catdb.RData"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="batch-go-term-enrichment-analysis">Batch GO term enrichment analysis</h3>
<p>Apply the enrichment analysis to the DEG sets obtained in the above differential expression analysis. Note, in the following example the <em>FDR</em> filter is set here to an unreasonably high value, simply because of the small size of the toy data set used in this vignette. Batch enrichment analysis of many gene sets is performed with the <em><code class="highlighter-rouge">GOCluster_Report</code></em> function. When <em><code class="highlighter-rouge">method="all"</code></em>, it returns all GO terms passing the p-value cutoff specified under the <em><code class="highlighter-rouge">cutoff</code></em> arguments. When <em><code class="highlighter-rouge">method="slim"</code></em>, it returns only the GO terms specified under the <em><code class="highlighter-rouge">myslimv</code></em> argument. The given example shows how one can obtain such a GO slim vector from BioMart for a specific organism.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">load</span><span class="p">(</span><span class="s2">"data/GO/catdb.RData"</span><span class="p">)</span><span class="w">
</span><span class="n">DEG_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">edgeDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">50</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">up_down</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">UporDown</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">up_down</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">up_down</span><span class="p">),</span><span class="w"> </span><span class="s2">"_up_down"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">up</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Up</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">up</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">up</span><span class="p">),</span><span class="w"> </span><span class="s2">"_up"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">down</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Down</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">down</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">down</span><span class="p">),</span><span class="w"> </span><span class="s2">"_down"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">DEGlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">up_down</span><span class="p">,</span><span class="w"> </span><span class="n">up</span><span class="p">,</span><span class="w"> </span><span class="n">down</span><span class="p">)</span><span class="w">
</span><span class="n">DEGlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEGlist</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">]</span><span class="w">
</span><span class="n">BatchResult</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GOCluster_Report</span><span class="p">(</span><span class="n">catdb</span><span class="o">=</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">setlist</span><span class="o">=</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"all"</span><span class="p">,</span><span class="w"> </span><span class="n">id_type</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">,</span><span class="w"> </span><span class="n">CLSZ</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">gocats</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"MF"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CC"</span><span class="p">),</span><span class="w"> </span><span class="n">recordSpecGO</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="s2">"biomaRt"</span><span class="p">);</span><span class="w"> </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="o">=</span><span class="s2">"athaliana_eg_gene"</span><span class="p">)</span><span class="w">
</span><span class="n">goslimvec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"goslim_goa_accession"</span><span class="p">),</span><span class="w"> </span><span class="n">mart</span><span class="o">=</span><span class="n">m</span><span class="p">)[,</span><span class="m">1</span><span class="p">])</span><span class="w">
</span><span class="n">BatchResultslim</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GOCluster_Report</span><span class="p">(</span><span class="n">catdb</span><span class="o">=</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">setlist</span><span class="o">=</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"slim"</span><span class="p">,</span><span class="w"> </span><span class="n">id_type</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">,</span><span class="w"> </span><span class="n">myslimv</span><span class="o">=</span><span class="n">goslimvec</span><span class="p">,</span><span class="w"> </span><span class="n">CLSZ</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="n">gocats</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"MF"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CC"</span><span class="p">),</span><span class="w"> </span><span class="n">recordSpecGO</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="plot-batch-go-term-results">Plot batch GO term results</h3>
<p>The <em><code class="highlighter-rouge">data.frame</code></em> generated by <em><code class="highlighter-rouge">GOCluster_Report</code></em> can be plotted with the <em><code class="highlighter-rouge">goBarplot</code></em> function. Because of the variable size of the sample sets, it may not always be desirable to show the results from different DEG sets in the same bar plot. Plotting single sample sets is achieved by subsetting the input data frame as shown in the first line of the following example.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"M6-V6_up_down"</span><span class="p">,</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="o">$</span><span class="n">CLID</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">gos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"GOslimbarplotMF.pdf"</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">10</span><span class="p">);</span><span class="w"> </span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"MF"</span><span class="p">);</span><span class="w"> </span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"BP"</span><span class="p">)</span><span class="w">
</span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"CC"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<p><img src="GOslimbarplotMF.png" alt="" /></p>
<div align="center">**Figure 7:** GO Slim Barplot for MF Ontology.</div>
<p>&lt;/br&gt;</p>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="clustering-and-heat-maps">Clustering and heat maps</h2>
<p>The following example performs hierarchical clustering on the <em><code class="highlighter-rouge">rlog</code></em> transformed expression matrix subsetted by the DEGs identified in the 
above differential expression analysis. It uses a Pearson correlation-based distance measure and complete linkage for cluster joining.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span><span class="w">
</span><span class="n">geneids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">DEG_list</span><span class="p">[[</span><span class="m">1</span><span class="p">]])))</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">assay</span><span class="p">(</span><span class="n">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">))[</span><span class="n">geneids</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"heatmap1.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="o">=</span><span class="s2">"row"</span><span class="p">,</span><span class="w"> </span><span class="n">clustering_distance_rows</span><span class="o">=</span><span class="s2">"correlation"</span><span class="p">,</span><span class="w"> </span><span class="n">clustering_distance_cols</span><span class="o">=</span><span class="s2">"correlation"</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>
<p><img src="heatmap1.png" alt="" /></p>
<div align="center">**Figure 8:** Heat map with hierarchical clustering dendrograms of DEGs.</div>
<p>&lt;/br&gt;</p>

<div align="right">[Back to Table of Contents]()</div>

<h1 id="workflow-templates">Workflow templates</h1>

<h2 id="rna-seq-sample">RNA-Seq sample</h2>

<p>Load the RNA-Seq sample workflow into your current working directory.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="run-workflow">Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code class="highlighter-rouge">systemPipeRNAseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/rnaseq/systemPipeRNAseq_single.Rnw">Rnw</a>) by executing from the command-line <em><code class="highlighter-rouge">make -B</code></em> within the <em><code class="highlighter-rouge">rnaseq</code></em> directory. Alternatively, one can run the code from the provided <em><code class="highlighter-rouge">*.Rnw</code></em> template file from within R interactively.</p>

<p>Workflow includes following steps:</p>

<ol>
  <li>Read preprocessing
    <ul>
      <li>Quality filtering (trimming)</li>
      <li>FASTQ quality report</li>
    </ul>
  </li>
  <li>Alignments: <em><code class="highlighter-rouge">Tophat2</code></em> (or any other RNA-Seq aligner)</li>
  <li>Alignment stats</li>
  <li>Read counting</li>
  <li>Sample-wise correlation analysis</li>
  <li>Analysis of differentially expressed genes (DEGs)</li>
  <li>GO term enrichment analysis</li>
  <li>Gene-wise clustering</li>
</ol>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="chip-seq-sample">ChIP-Seq sample</h2>

<p>Load the ChIP-Seq sample workflow into your current working directory.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"chipseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"chipseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="run-workflow-1">Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code class="highlighter-rouge">systemPipeChIPseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/chipseq/systemPipeChIPseq_single.Rnw">Rnw</a>) by executing from the command-line <em><code class="highlighter-rouge">make -B</code></em> within the <em><code class="highlighter-rouge">chipseq</code></em> directory. Alternatively, one can run the code from the provided <em><code class="highlighter-rouge">*.Rnw</code></em> template file from within R interactively.</p>

<p>Workflow includes following steps:</p>

<ol>
  <li>Read preprocessing
    <ul>
      <li>Quality filtering (trimming)</li>
      <li>FASTQ quality report</li>
    </ul>
  </li>
  <li>Alignments: <em><code class="highlighter-rouge">Bowtie2</code></em> or <em><code class="highlighter-rouge">rsubread</code></em></li>
  <li>Alignment stats</li>
  <li>Peak calling: <em><code class="highlighter-rouge">MACS2</code></em>, <em><code class="highlighter-rouge">BayesPeak</code></em></li>
  <li>Peak annotation with genomic context</li>
  <li>Differential binding analysis</li>
  <li>GO term enrichment analysis</li>
  <li>Motif analysis</li>
</ol>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="var-seq-sample">VAR-Seq sample</h2>

<h3 id="var-seq-workflow-for-single-machine">VAR-Seq workflow for single machine</h3>

<p>Load the VAR-Seq sample workflow into your current working directory.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"varseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"varseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="run-workflow-2">Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code class="highlighter-rouge">systemPipeVARseq_single</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq_single.Rnw">Rnw</a>) by executing from the command-line <em><code class="highlighter-rouge">make -B</code></em> within the <em><code class="highlighter-rouge">varseq</code></em> directory. Alternatively, one can run the code from the provided <em><code class="highlighter-rouge">*.Rnw</code></em> template file from within R interactively.</p>

<p>Workflow includes following steps:</p>

<ol>
  <li>Read preprocessing
    <ul>
      <li>Quality filtering (trimming)</li>
      <li>FASTQ quality report</li>
    </ul>
  </li>
  <li>Alignments: <em><code class="highlighter-rouge">gsnap</code></em>, <em><code class="highlighter-rouge">bwa</code></em></li>
  <li>Variant calling: <em><code class="highlighter-rouge">VariantTools</code></em>, <em><code class="highlighter-rouge">GATK</code></em>, <em><code class="highlighter-rouge">BCFtools</code></em></li>
  <li>Variant filtering: <em><code class="highlighter-rouge">VariantTools</code></em> and <em><code class="highlighter-rouge">VariantAnnotation</code></em></li>
  <li>Variant annotation: <em><code class="highlighter-rouge">VariantAnnotation</code></em></li>
  <li>Combine results from many samples</li>
  <li>Summary statistics of samples</li>
</ol>

<div align="right">[Back to Table of Contents]()</div>

<h3 id="var-seq-workflow-for-computer-cluster">VAR-Seq workflow for computer cluster</h3>
<p>The workflow template provided for this step is called <em><code class="highlighter-rouge">systemPipeVARseq.Rnw</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/varseq/systemPipeVARseq.Rnw">Rnw</a>).
It runs the above VAR-Seq workflow in parallel on multiple computer nodes of an HPC system using Torque as scheduler.</p>

<div align="right">[Back to Table of Contents]()</div>

<h2 id="ribo-seq-sample">Ribo-Seq sample</h2>

<p>Load the Ribo-Seq sample workflow into your current working directory.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"riboseq"</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"riboseq"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h3 id="run-workflow-3">Run workflow</h3>
<p>Next, run the chosen sample workflow <em><code class="highlighter-rouge">systemPipeRIBOseq</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/riboseq/systemPipeRIBOseq.pdf?raw=true">PDF</a>, <a href="https://github.com/tgirke/systemPipeRdata/blob/master/inst/extdata/workflows/ribseq/systemPipeRIBOseq_single.Rnw">Rnw</a>) by executing from the command-line <em><code class="highlighter-rouge">make -B</code></em> within the <em><code class="highlighter-rouge">ribseq</code></em> directory. Alternatively, one can run the code from the provided <em><code class="highlighter-rouge">*.Rnw</code></em> template file from within R interactively.</p>

<p>Workflow includes following steps:</p>

<ol>
  <li>Read preprocessing
    <ul>
      <li>Adaptor trimming and quality filtering</li>
      <li>FASTQ quality report</li>
    </ul>
  </li>
  <li>Alignments: <em><code class="highlighter-rouge">Tophat2</code></em> (or any other RNA-Seq aligner)</li>
  <li>Alignment stats</li>
  <li>Compute read distribution across genomic features</li>
  <li>Adding custom features to workflow (e.g. uORFs)</li>
  <li>Genomic read coverage along transcripts</li>
  <li>Read counting</li>
  <li>Sample-wise correlation analysis</li>
  <li>Analysis of differentially expressed genes (DEGs)</li>
  <li>GO term enrichment analysis</li>
  <li>Gene-wise clustering</li>
  <li>Differential ribosome binding (translational efficiency)</li>
</ol>

<div align="right">[Back to Table of Contents]()</div>

<h1 id="version-information">Version information</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.2.3 (2015-12-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.3 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  utils     datasets  grDevices methods   base     
## 
## other attached packages:
##  [1] DESeq2_1.10.0              RcppArmadillo_0.6.300.2.0  Rcpp_0.12.3               
##  [4] ape_3.4                    ggplot2_2.0.0              systemPipeR_1.4.8         
##  [7] RSQLite_1.0.0              DBI_0.3.1                  ShortRead_1.28.0          
## [10] GenomicAlignments_1.6.1    SummarizedExperiment_1.0.1 Biobase_2.30.0            
## [13] BiocParallel_1.4.0         Rsamtools_1.22.0           Biostrings_2.38.2         
## [16] XVector_0.10.0             GenomicRanges_1.22.1       GenomeInfoDb_1.6.1        
## [19] IRanges_2.4.4              S4Vectors_0.8.3            BiocGenerics_0.16.1       
## [22] BiocStyle_1.8.0           
## 
## loaded via a namespace (and not attached):
##  [1] edgeR_3.12.0           splines_3.2.3          Formula_1.2-1          latticeExtra_0.6-26   
##  [5] RBGL_1.46.0            yaml_2.1.13            Category_2.36.0        lattice_0.20-33       
##  [9] limma_3.26.3           digest_0.6.9           RColorBrewer_1.1-2     checkmate_1.6.3       
## [13] colorspace_1.2-6       htmltools_0.3          Matrix_1.2-3           plyr_1.8.3            
## [17] GSEABase_1.32.0        XML_3.98-1.3           pheatmap_1.0.7         biomaRt_2.26.1        
## [21] genefilter_1.52.0      zlibbioc_1.16.0        xtable_1.8-0           GO.db_3.2.2           
## [25] scales_0.3.0           brew_1.0-6             annotate_1.48.0        GenomicFeatures_1.22.6
## [29] nnet_7.3-11            proto_0.3-10           survival_2.38-3        magrittr_1.5          
## [33] evaluate_0.8           fail_1.3               nlme_3.1-122           foreign_0.8-66        
## [37] hwriter_1.3.2          GOstats_2.36.0         graph_1.48.0           tools_3.2.3           
## [41] formatR_1.2.1          BBmisc_1.9             stringr_1.0.0          sendmailR_1.2-1       
## [45] locfit_1.5-9.1         munsell_0.4.2          cluster_2.0.3          AnnotationDbi_1.32.0  
## [49] lambda.r_1.1.7         futile.logger_1.4.1    grid_3.2.3             RCurl_1.95-4.7        
## [53] rjson_0.2.15           AnnotationForge_1.12.0 labeling_0.3           bitops_1.0-6          
## [57] base64enc_0.1-3        rmarkdown_0.9.2        gtable_0.1.2           codetools_0.2-14      
## [61] gridExtra_2.0.0        knitr_1.12             rtracklayer_1.30.1     Hmisc_3.17-0          
## [65] futile.options_1.0.0   stringi_1.0-1          BatchJobs_1.6          geneplotter_1.48.0    
## [69] rpart_4.1-10           acepack_1.3-3.3
</code></pre>
</div>
<div align="right">[Back to Table of Contents]()</div>

<h1 id="references">References</h1>


<div class="tags">
    
</div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 Thomas Girke. All rights reserved. <br />
 Site last generated: Jul 4, 2016 <br />
<p><img src="/manuals/images/company_logo.png" alt="Company logo"/></p>
                </div>
            </div>
</footer>

    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>
    <script>
    jQuery(function ($) {
        $('.post-content').annotator()
                .annotator('setupPlugins', {tokenUrl: 'http://example.com/api/token'})
    });
    </script>
</body>

</html>